x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -410,6 +449,21 @@ public void toXContentFragment(XContentBuilder builder) throws IOException {
         }
     }
 
+    private void initializeSubjects() {
+        if (authenticatingSubject == null) {
+            if (user.isRunAs()) {
+                authenticatingSubject = new Subject(user.authenticatedUser(), authenticatedBy, version, metadata);
+                // The lookup user for run-as currently don't have authentication metadata associated with them because
+                // lookupUser only returns the User object. The lookup user for authorization delegation does have
+                // authentication metadata, but the realm does not expose this difference between authenticatingUser and
+                // delegateUser so effectively this is handled together with the authenticatingSubject not effectiveSubject.
+                effectiveSubject = new Subject(user, lookedUpBy, version, Map.of());
+            } else {
+                authenticatingSubject = effectiveSubject = new Subject(user, authenticatedBy, version, metadata);
+            }
+        }
```

Comment:
On a second thought, this method needs to be thread-safe because it's setting two variables so inherently not atomic. 

Taking a step back, I was trying to save some computation by delay the initialising the subjects. But I think the save here is too marginal. Introducing a synchornisation is likely to defeat the purpose. So I just inlined it into the constructors.